docs/manpages/tracker-processes.1
.TH tracker-processes 1 "September 2009" GNU "User Commands"

.SH NAME
tracker-processes \- Manage Tracker processes and data

.SH SYNOPSIS
\fBtracker-processes\fR [\fIOPTION\fR...]

.SH DESCRIPTION
.B tracker-processes
lists all tracker related processes when no
\fIOPTION\fRs are specified.

.SH OPTIONS
.TP
.B \-?, \-\-help
Show summary of options.
.TP
.B \-k, \-\-kill
This uses SIGKILL to stop all Tracker processes found. This is not
advised unless you are having problems stopping Tracker in the first
place. This 
.B GUARANTEES
death.
.TP
.B \-t, \-\-terminate
This uses SIGTERM to stop all Tracker processes found. This is
recommended over \-\-kill because it gives the processes time to
shutdown cleanly.
.TP
.B \-s, \-\-soft-reset
A soft reset works exactly the same way that 
.B \-\-hard-reset 
does, with the exception that the backup and journal are not removed.
These are restored when 
.B tracker-store
is restarted.

.TP
.B \-r, \-\-hard-reset
This kills all processes in the same way that 
.B \-\-kill
does but it also removes all databases. Restarting 
.B tracker-store
re-creates the databases.

.SH SEE ALSO
.BR tracker-store (1).


NEWS
NEW in 0.7.1
============

Improvements / New:

  General:
  * Updated version of dbus and dbus-glib and removed
    SUBJECT_TO_CHANGE use everywhere.
  * Updated version of libxml to something realistic.
  * Updated version of libgee to 0.3.
  * Updated vala requirement suggestion in autogen.sh to 0.7.6.
  
  Ontology:
  * Added prefix "mlo" for http://www.tracker-project.org/temp/mlo
  * Updated scal ontology to cover ical standard.

  Documentation: 
  * Added initial documentation for libtracker-miner.
  * Fixed typo in tracker-processes man page.
  * Fixed warnings when building documenation for libtracker-common
  * Regenerated documentation sections for libtracker-common

  libtracker-common:
  * Fixed memory leak in tracker_string_to_date().
  * Moved TrackerParser API to libtracker-fts, only used there.

  libtracker-db:
  * Updated headers so only one header can be included if used externally

  libtracker-data:
  * Added support for fn:contains and fn:ends-with for SPARQL.
  * Removed unnecessary FTS INSERT and DELETE statements.
  * Updated to avoid unnecessary SQL CAST expressions.

  libtracker-fts:
  * Added to unit tests
  * Removed broken and unused paragraph detection.
  * Moved this from "tracker-fts" everywhere in code base since it is
    a library not an application.

  libtracker-miner:
  * Added get_status() and is_paused() APIs.
  * Removed GetName and GetDescription DBus methods, now in desktop
    files and exposed through TrackerMinerManager API.

  tracker-store:
  * Clean up queued batch updates to the database when clients
    disconnect.
  * Do not use JOIN for simple OPTIONAL patterns in SPARQL
    (significant speed improvement).

  tracker-miner-fs:
  * Added support back for thumbnailing.
  * Added support for full paths in config for IgnoredFiles and
    IgnoredDirectories.
  * Added support for XDG spec. for special user directories. So now
    in the config, IndexRecursiveDirectories and
    IndexSingleDirectories can include "&DOCUMENTS", or "&MUSIC", etc. 
    The list of special locations is listed in the config and
    documented. This falls back to $HOME when not available for some
    distributions.
  * Removed unused files.
  * Updated the default indexed locations to DESKTOP, DOCUMENTS,
    DOWNLOADS, MUSIC, PICTURES and VIDEOS as per the XDG spec (this
    was $HOME).
  * Fixed trying to send thumbnail queues of size 0.
  * Fixed trying to send thumbnails when service is not available.
  * Fixed double unref in thumbnailer code.
  * Fixed 2 memory leaks found while valgrinding.

  tracker-extract:
  * Fixed assertion in HAL is NULL from starting on the command line.
  * Fixed Ctrl+C not working with -d.
    
  tracker-status:
  * Use libtracker-miner APIs instead of DBus calls.
  * Use libtracker-miner API to get display names (translatable).

  tracker-search-bar:
  * Fixed crash when clicking on the results window with the "no
    results found..." message.

  tracker-search-tool:
  * Added smart wildcards when last character is not a space.
  * Fixed category selector.
  * Fixed crasher when selecting categories.

Bugs:

  * Fixes NB#141309, Remove nco:hasPostalAddress cardinality limitation
  * Fixes NB#132655, Tracker-import error handling jams
  * Fixes GB#595976, Message 'SQL logic error or missing database'
  * Fixes GB#592400, String translation issues
  * Fixes GB#595752, Minor string fixes

Notes:

  The Deskbar applet packaged with Tracker is currently out of date
  and needs some love to get it working with the new APIs. It is in
  the same state as it was for 0.6. As such it isn't expected to work.


NEW in 0.7
==========

Improvements / New:

  * Library: libtracker-miner
    A common library for data miners to use allowing easy integration
    with Tracker. This provides D-Bus integration, state and progress
    APIs and a means to pause / resume your miner.

  * Library: libtracker-client
    This API has been completely changed. It is much simpler now and
    relies more on the SPARQL language than specific APIs to query
    Tracker. 

  * Daemon: tracker-store
    Completely reworked to accept SPARQL queries and updates.
    Implements Nepomuk ontology instead of flawed own brand. This
    ontology is implemented following the W3C RDF semantics.

  * Database: Completely redesigned ALL databases.
    These now support triple store and allow much faster querying.

  * Application: tracker-miner-fs
    Implements libtracker-miner and facilitates the old roles of
    tracker-indexer and some of the old roles from trackerd like file
    system crawling. This mines data and sends it to tracker-store to
    put in the database about files AND applications.

  * Application: tracker-sparql
    A command line tool to query or update the databases using SPARQL.

  * Application: tracker-import
    A command line tool to import data kept in Turtle files.

  * Application: tracker-search-bar
    An applet for quickly searching for music, videos, images,
    documents and folders.

  * Application: tracker-status-icon
    Previously tracker-applet, but rewritten and now much simpler.
    This provides state and progress information about all miners
    which are running. This also allows the user to pause / resume
    data miners.

  * Application: tracker-search-tool
    This has been completely rewritten using vala and the new
    architecture underneath it. Support for the old features are not
    quite there yet, but it is similar to its predecessor!

  * Application: tracker-explore
    This is an application written to explore data in the database and
    each item's relationship to the next. This really demonstrates
    Tracker's ontology and provides it through full text searching.

  * Plugin: Evolution
    Now the evolution plugin is supported by the developers instead of
    us dipping into their databases (as previously) which is dangerous
    and unsafe. Also this implements the libtracker-miner API so it

src/tracker-utils/tracker-processes.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/*
 * Copyright (C) 2006, Mr Jamie McCracken (jamiemcc@gnome.org)
 * Copyright (C) 2008, Nokia
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#include "config.h"

#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <locale.h>
#include <errno.h>
#include <sys/types.h>
#include <signal.h>

#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gprintf.h>

#include <libtracker-common/tracker-common.h>

#include <libtracker-db/tracker-db-manager.h>

static gboolean     should_kill;
static gboolean     should_terminate;
static gboolean     hard_reset, soft_reset;

static GOptionEntry entries[] = {
	{ "kill", 'k', 0, G_OPTION_ARG_NONE, &should_kill,
	  N_("Use SIGKILL to stop all tracker processes found - guarantees death :)"),
	  NULL },
	{ "terminate", 't', 0, G_OPTION_ARG_NONE, &should_terminate,
	  N_("Use SIGTERM to stop all tracker processes found"),
	  NULL 
	},
	{ "soft-reset", 's', 0, G_OPTION_ARG_NONE, &soft_reset,
	  N_("This will kill all Tracker processes and remove all databases except the backup and journal (a restart will restore the data)"),
	  NULL },
	{ "hard-reset", 'r', 0, G_OPTION_ARG_NONE, &hard_reset,
	  N_("This will kill all Tracker processes and remove all databases"),
	  NULL },
	{ NULL }
};

static GSList *
get_pids (void)
{
	GError *error = NULL;
	GDir *dir;
	GSList *pids = NULL;
	const gchar *name;

	dir = g_dir_open ("/proc", 0, &error);
	if (error) {
		g_printerr ("%s, %s\n",
			    _("Could not open /proc"),
			    error ? error->message : _("no error given"));
		g_clear_error (&error);
		return NULL;
	}

	while ((name = g_dir_read_name (dir)) != NULL) { 
		gchar c;
		gboolean is_pid = TRUE;

		for (c = *name; c && c != ':' && is_pid; c++) {		
			is_pid &= g_ascii_isdigit (c);
		}

		if (!is_pid) {
			continue;
		}

		pids = g_slist_prepend (pids, g_strdup (name));
	}

	g_dir_close (dir);

	return g_slist_reverse (pids);
}

static void
log_handler (const gchar    *domain,
	     GLogLevelFlags  log_level,
	     const gchar    *message,
	     gpointer	     user_data)
{
	switch (log_level) {
	case G_LOG_LEVEL_WARNING:
	case G_LOG_LEVEL_CRITICAL:
	case G_LOG_LEVEL_ERROR:
	case G_LOG_FLAG_RECURSION:
	case G_LOG_FLAG_FATAL:
		g_fprintf (stderr, "%s\n", message);
		fflush (stderr);
		break;
	case G_LOG_LEVEL_MESSAGE:
	case G_LOG_LEVEL_INFO:
	case G_LOG_LEVEL_DEBUG:
	case G_LOG_LEVEL_MASK:
	default:
		g_fprintf (stdout, "%s\n", message);
		fflush (stdout);
		break;
	}	
}

int
main (int argc, char **argv)
{
	GOptionContext *context;
	GError *error = NULL;
	GSList *pids;
	GSList *l;
	gchar  *str;

	setlocale (LC_ALL, "");

	bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	g_type_init ();

	/* Translators: this messagge will apper immediately after the	*/
	/* usage string - Usage: COMMAND [OPTION]... <THIS_MESSAGE>	*/
	context = g_option_context_new (_(" - Manage Tracker processes and data"));
	g_option_context_add_main_entries (context, entries, NULL);
	g_option_context_parse (context, &argc, &argv, NULL);
	g_option_context_free (context);

	if (should_kill && should_terminate) {
		g_printerr ("%s\n",
			    _("You can not use the --kill and --terminate arguments together"));
		return EXIT_FAILURE;
	} else if ((hard_reset || soft_reset) && should_terminate) {
		g_printerr ("%s\n",
			    _("You can not use the --terminate with --hard-reset, --kill is implied"));
		return EXIT_FAILURE;
	}

	if (hard_reset || soft_reset) {
		/* Imply --kill */
		should_kill = TRUE;
	}

	pids = get_pids ();
	str = g_strdup_printf (tracker_dngettext (NULL,
						  "Found %d PID…", 
						  "Found %d PIDs…",
						  g_slist_length (pids)),
			       g_slist_length (pids));
	g_print ("%s\n", str);
	g_free (str);

	for (l = pids; l; l = l->next) {
		gchar *filename;
		gchar *contents = NULL;
		gchar **strv;

		filename = g_build_filename ("/proc", l->data, "cmdline", NULL);
		if (!g_file_get_contents (filename, &contents, NULL, &error)) {
			str = g_strdup_printf (_("Could not open '%s'"), filename);
			g_printerr ("%s, %s\n", 
				    str,
				    error ? error->message : _("no error given"));
			g_free (str);
			g_clear_error (&error);
			g_free (contents);
			g_free (filename);

			continue;
		}
		
		strv = g_strsplit (contents, "^@", 2);
		if (strv && strv[0]) {
			gchar *basename;

			basename = g_path_get_basename (strv[0]);
			if (g_str_has_prefix (basename, "tracker") == TRUE &&
			    g_str_has_suffix (basename, "-processes") == FALSE) {
				pid_t pid;

				pid = atoi (l->data);
				str = g_strdup_printf (_("Found process ID %d for '%s'"), pid, basename);
				g_print ("%s\n", str);
				g_free (str);

				if (should_terminate) {
					if (kill (pid, SIGTERM) == -1) {
						const gchar *errstr = g_strerror (errno);
						
						str = g_strdup_printf (_("Could not terminate process %d"), pid);
						g_printerr ("  %s, %s\n", 
							    str,
							    errstr ? errstr : _("no error given"));
						g_free (str);
					} else {
						str = g_strdup_printf (_("Terminated process %d"), pid);
						g_print ("  %s\n", str);
						g_free (str);
					}
				} else if (should_kill) {
					if (kill (pid, SIGKILL) == -1) {
						const gchar *errstr = g_strerror (errno);
						
						str = g_strdup_printf (_("Could not kill process %d"), pid);
						g_printerr ("  %s, %s\n", 
							    str,
							    errstr ? errstr : _("no error given"));
						g_free (str);
					} else {
						str = g_strdup_printf (_("Killed process %d"), pid);
						g_print ("  %s\n", str);
						g_free (str);
					}
				}
			}

			g_free (basename);
		}

		g_strfreev (strv);
		g_free (contents);
		g_free (filename);
	}

	g_slist_foreach (pids, (GFunc) g_free, NULL);
	g_slist_free (pids);

	if (hard_reset || soft_reset) {
		guint log_handler_id;

		/* Set log handler for library messages */
		log_handler_id = g_log_set_handler (NULL,
						    G_LOG_LEVEL_MASK | G_LOG_FLAG_FATAL,
						    log_handler,
						    NULL);
		
		g_log_set_default_handler (log_handler, NULL);

		/* Clean up */
		if (!tracker_db_manager_init (TRACKER_DB_MANAGER_REMOVE_ALL, NULL, FALSE, NULL)) {
			return EXIT_FAILURE;
		}

		tracker_db_manager_remove_all (hard_reset);
		tracker_db_manager_shutdown ();

		/* Unset log handler */
		g_log_remove_handler (NULL, log_handler_id);
	}

	return EXIT_SUCCESS;
}
